diffrax/_brownian/tree.py

```diff
@@ -72,55 +92,72 @@ class _State(eqx.Module):
     bkk_s_u_su: Optional[FloatTriple]  # \bar{K}_s, _u, _{s,u}
 
 
-def _levy_diff(_, x0: tuple, x1: tuple) -> AbstractBrownianIncrement:
+def _levy_diff(_, x0: _LevyVal, x1: _LevyVal) -> AbstractBrownianIncrement:
     r"""Computes $(W_{s,u}, H_{s,u})$ from $(W_s, \bar{H}_{s,u})$ and
     $(W_u, \bar{H}_u)$, where $\bar{H}_u = u * H_u$.
 
     **Arguments:**
 
     - `_`: unused, for the purposes of aligning the `jtu.tree_map`.
-    - `x0`: `LevyVal` at time `s`.
-    - `x1`: `LevyVal` at time `u`.
+    - `x0`: `_AbstractLevyVal` at time `s`.
+    - `x1`: `_AbstractLevyVal` at time `u`.
 
     **Returns:**
 
-    `LevyVal(W_su, H_su)`
+    `AbstractBrownianIncrement(W_su, H_su, K_su)`
     """
+    dtype = jnp.dtype(x0.W)
+    tdtype = complex_to_real_dtype(dtype)
+    su = jnp.asarray(x1.dt - x0.dt, dtype=tdtype)
+    if x0.H is None:  # BM only case
+        assert x1.H is None
+        return BrownianIncrement(dt=su, W=x1.W - x0.W)
+
+    # the following computation is common to the space-time
+    # and the space-time-time Levy area case
+    assert x0.H is not None
+    assert x1.H is not None
+    assert x0.bar_H is not None
+    assert x1.bar_H is not None
+    w_su = x1.W - x0.W
+    inverse_su = 1 / jnp.where(jnp.abs(su) < jnp.finfo(su).eps, jnp.inf, su)
+    with jax.numpy_dtype_promotion("standard"):
+        u_bb_s = x1.dt * x0.W - x0.dt * x1.W
+        bhh_su = x1.bar_H - x0.bar_H - 0.5 * u_bb_s  # bhh_su = H_{s,u} * (u-s)
+        hh_su = inverse_su * bhh_su
+
+    if x0.K is None:  # space-time Levy area case
+        return SpaceTimeLevyArea(dt=su, W=w_su, H=hh_su)
 
-    if len(x0) == 2:  # BM only case
-        assert len(x1) == 2
-        dt0, w0 = x0
-        dt1, w1 = x1
-        su = jnp.asarray(dt1 - dt0, dtype=complex_to_real_dtype(w0.dtype))
-        return BrownianIncrement(dt=su, W=w1 - w0)
-
-    elif len(x0) == 4:  # space-time levy area case
-        assert len(x1) == 4
-        dt0, w0, hh0, bhh0 = x0
-        dt1, w1, hh1, bhh1 = x1
-
-        w_su = w1 - w0
-        su = jnp.asarray(dt1 - dt0, dtype=complex_to_real_dtype(w0.dtype))
-        _su = jnp.where(jnp.abs(su) < jnp.finfo(su).eps, jnp.inf, su)
-        inverse_su = 1 / _su
+    elif x0.K is not None:  # space-time-time Levy area case
+        assert x1.K is not None
+        assert x0.bar_K is not None
+        assert x1.bar_K is not None
         with jax.numpy_dtype_promotion("standard"):
-            u_bb_s = dt1 * w0 - dt0 * w1
-            bhh_su = bhh1 - bhh0 - 0.5 * u_bb_s  # bhh_su = H_{s,u} * (u-s)
-            hh_su = inverse_su * bhh_su
-        return SpaceTimeLevyArea(dt=su, W=w_su, H=hh_su)
+            bkk_su = (
+                x1.bar_K
+                - x0.bar_K
+                - su / 2 * x0.bar_H
+                + x0.dt / 2 * bhh_su
+                - (x1.dt - 2 * x0.dt) / 12 * u_bb_s
```

Comment:
Are these meant to be e.g. `su / (2 * x0.bar_H)` or `(su / 2) * x0.bar_H`?

I usually prefer to add brackets to communicate intent, even if the language has the appropriate binding precedences to make them unnecessary.

Reply:
I meant the latter, but yes, fair point.